### PR TITLE
net: openthread: rpc: diagnostic functions implementation.

### PR DESCRIPTION
**Pull request by canisLupus1313**
Implementation of RCP diagnostic api set.

### CODE REVIEW

**Comment by alxelax** [comment]
@canisLupus1313, could you fix compliance issues? It is difficult to review file that is polluted by CI messages. :)

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    struct zcbor_string zst;
+	struct nrf_rpc_cbor_ctx ctx;
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, cbor_buffer_size);
+
+	nrf_rpc_cbor_cmd_rsp(&ot_group, id, &ctx);
+
+    zcbor_bstr_decode(ctx.zs, &zst);
+	nrf_rpc_cbor_decoding_done(&ot_group, &ctx);
+	return zst;
```
Comment (Damian-Nordic) [inline subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:50]: This returns a pointer to a buffer that is deallocated in `_decoding_done`.

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+const otMacCounters *otLinkGetCounters(otInstance *aInstance) {
+	return 0;
+}
+
+const otExtAddress *otLinkGetExtendedAddress(otInstance *aInstance) {
+	struct zcbor_string zst = get_string(OT_RPC_CMD_IF_EXTADDR);
+
+	memcpy(mac.m8, zst.value, sizeof(mac) < zst.len ? sizeof(mac) : zst.len);
```
Comment (Damian-Nordic) [inline subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:78]: I think this should validate the length and report an error instead of dropping the superfluous bytes.

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+#include <openthread/link.h>
+#include <openthread/thread.h>
+#include <sys/_stdint.h>
+
+NRF_RPC_GROUP_DECLARE(ot_group);
+
+static char version[256]; //to validate
+static otExtAddress mac = {.m8 = {0xF4, 0xCE, 0x36, 0x24, 0x95, 0x6A, 0xD1, 0xD0}};
+static otExtendedPanId ext_pan_id;
+static char thread_network_name[256];
```
Comment (Damian-Nordic) [inline subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:23]:
```suggestion
static char thread_network_name[OT_NETWORK_NAME_MAX_SIZE +1];
```

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+#include <openthread/dataset.h>
+#include <openthread/instance.h>
+#include <openthread/link.h>
+#include <openthread/thread.h>
+#include <sys/_stdint.h>
+
+NRF_RPC_GROUP_DECLARE(ot_group);
+
+static char version[256]; //to validate
+static otExtAddress mac = {.m8 = {0xF4, 0xCE, 0x36, 0x24, 0x95, 0x6A, 0xD1, 0xD0}};
```
Comment (Damian-Nordic) [inline subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:21]: Why the default?

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+static void get_uint_t(ot_rpc_id id, void *result, size_t result_size) {
+	const size_t cbor_buffer_size = 0;
+	struct nrf_rpc_cbor_ctx ctx;
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, cbor_buffer_size);
+
+	nrf_rpc_cbor_cmd_rsp(&ot_group, id, &ctx);
+
+    zcbor_uint_decode(ctx.zs, result, result_size);
```
Comment (Damian-Nordic) [inline subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:34]: We should handle the decode error somehow (e.g. by calling nrf_rpc_err() and returning 0).

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+NRF_RPC_GROUP_DECLARE(ot_group);
+
+static char version[256]; //to validate
+static otExtAddress mac = {.m8 = {0xF4, 0xCE, 0x36, 0x24, 0x95, 0x6A, 0xD1, 0xD0}};
+static otExtendedPanId ext_pan_id;
+static char thread_network_name[256];
+static otMeshLocalPrefix mesh_prefix;
+
+static void get_uint_t(ot_rpc_id id, void *result, size_t result_size) {
```
Comment (Damian-Nordic) [inline subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:26]: Consider extracting these helpers into some `client/ot_rpc_common_getter.h` (and adding `ot_rpc_` prefix to the function names) so that we can reuse them in other command groups.

File: subsys/net/openthread/rpc/common/ot_rpc_ids.h [hunk 7]
```diff
@@ -6,15 +6,15 @@
 
 /** @brief Command IDs accepted by the OpenThread over RPC client.
  */
-enum ot_rpc_cmd_client {
+enum OT_RPC_CMD__rpc_cmd_client {
```
Comment (Damian-Nordic) [inline subsys/net/openthread/rpc/common/ot_rpc_ids.h:9]: This doesn't seem to follow naming conventions used elsewhere in NCS. Why this change?

File: subsys/net/openthread/rpc/common/ot_rpc_ids.h [hunk 8]
```diff
@@ -23,4 +23,23 @@ enum ot_rpc_cmd_server {
 	OT_RPC_CMD_IF_ENABLE,
 	OT_RPC_CMD_IF_SEND,
 	OT_RPC_CMD_IF_EXTADDR,
```
Comment (Damian-Nordic) [inline subsys/net/openthread/rpc/common/ot_rpc_ids.h:25]: Should this be removed?

**Review by Damian-Nordic — COMMENTED** [review]
Reviewed client part for now.

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c (hunk 4 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:21]: No reason it was useful in testing.

File: subsys/net/openthread/rpc/common/ot_rpc_ids.h (hunk 8 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/common/ot_rpc_ids.h:25]: This is still needed during startup.

Comment (canisLupus1313) [reply subsys/net/openthread/rpc/common/ot_rpc_ids.h:25]: yes but the I remained the same.

File: subsys/net/openthread/rpc/common/ot_rpc_ids.h (hunk 7 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/common/ot_rpc_ids.h:9]: I don't recall how it got there. Never the less it's fixed.

File: applications/protocols_serialization/client/src/main.c [hunk 9]
```diff
@@ -6,6 +6,9 @@
 
 #include <zephyr/logging/log.h>
 
+#include <openthread/instance.h>
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply applications/protocols_serialization/client/src/main.c]: yep :)

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+{
+	return 0;
+}
+
+const otExtAddress *otLinkGetExtendedAddress(otInstance *aInstance)
+{
+	int ret_size = get_string(OT_RPC_CMD_LINK_GET_EXTENDED_ADDRESS, mac.m8, sizeof(mac.m8));
+
+	if (ret_size != sizeof(mac.m8)) {
+		__ASSERT(false, "Received mac addr size is too short");
```
Comment (alxelax) [inline subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:97]: I guess there should be just regular handler with logging of error or warning.

If customer builds it with assertion and deploy it in the `field` the device will have security vulnerability and potential vector for attack over rpc.

File: subsys/net/openthread/rpc/common/ot_rpc_ids.h [hunk 11]
```diff
@@ -14,7 +14,7 @@ enum ot_rpc_cmd_client {
 
 /** @brief Command IDs accepted by the OpenThread over RPC server.
  */
-enum ot_rpc_cmd_server {
+typedef enum ot_rpc_cmd_server {
```
Comment (alxelax) [inline subsys/net/openthread/rpc/common/ot_rpc_ids.h:17]: Why should it be type?

File: subsys/net/openthread/rpc/server/ot_rpc_diag_server.c [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+		const char * : serialize_str,                                                      \
+		const otExtAddress * : serialize_extaddr,                                          \
+		const otMeshLocalPrefix * : serialize_mesh_local_prefix,                           \
+		const otExtendedPanId  *  : serialize_ext_panid)(Z, X)
+
+#define NRF_RPC_CBOR_CMD_STRING_GETTER(func_name, func_to_call, call_id, type, ptr_call)           \
+	static void func_name(const struct nrf_rpc_group *group, struct nrf_rpc_cbor_ctx *ctx,     \
+			      void *handler_data)                                                  \
+	{                                                                                          \
+		size_t length = 16;                                                                \
```
Comment (alxelax) [inline subsys/net/openthread/rpc/server/ot_rpc_diag_server.c:87]: Why is it 16?

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+
+#include <openthread/error.h>
+#include <openthread/dataset.h>
+#include <openthread/instance.h>
+#include <openthread/link.h>
+#include <openthread/thread.h>
+
+NRF_RPC_GROUP_DECLARE(ot_group);
+
+static char version[256];
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:20]: Version size is not constant and differs form compilation to compilation.

File: subsys/net/openthread/rpc/common/ot_rpc_ids.h (hunk 11 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/common/ot_rpc_ids.h:17]: https://github.com/nrfconnect/sdk-nrf/pull/16326/files#diff-8db7e3b76c5d84d7087e0da2235cd6bc59a0e6b6aac837367d7d1f7f2127d902R26  IHMO it's better to define a type def than utilize a `int` to store the value.

File: subsys/net/openthread/rpc/server/ot_rpc_diag_server.c (hunk 12 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/server/ot_rpc_diag_server.c:87]: This is the reason: https://github.com/nrfconnect/sdk-nrf/blob/main/samples/nrf_rpc/entropy_nrf53/cpuapp/src/entropy_ser.c#L19 so all in al no perticular one i can change it.

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+const char *otGetVersionString(void)
+{
+	get_string(OT_RPC_CMD_GET_VERSION_STRING, version, sizeof(version));
+
+	return version;
+}
+
+otError otDatasetGetActive(otInstance *aInstance, otOperationalDataset *aDataset)
+{
+	return 0;
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:75]: I will implement it in separate PR

File: subsys/net/openthread/rpc/server/ot_rpc_diag_server.c [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+
+NRF_RPC_CBOR_CMD_STRING_GETTER(ot_get_local_prefix, otThreadGetMeshLocalPrefix,
+			       OT_RPC_CMD_THREAD_GET_MESH_LOCAL_PREFIX, otMeshLocalPrefix,
+			       openthread_get_default_context()->instance);
+
+#define NRF_RPC_CBOR_CMD_UINT_GETTER(func_name, func_to_call, call_id, type)                       \
+	static void func_name(const struct nrf_rpc_group *group, struct nrf_rpc_cbor_ctx *ctx,     \
+			      void *handler_data)                                                  \
+	{                                                                                          \
+		size_t length = 16;                                                                \
```
Comment (Damian-Nordic) [inline subsys/net/openthread/rpc/server/ot_rpc_diag_server.c:128]:
```suggestion
		size_t length = 1;                                                                \
```
should be enough

**Review by Damian-Nordic — APPROVED** [review]
final suggestions but lgtm overall

File: subsys/net/openthread/rpc/server/ot_rpc_diag_server.c (hunk 12 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/server/ot_rpc_diag_server.c:87]: Changed to @Damian-Nordic suggestion.

File: subsys/net/openthread/rpc/server/ot_rpc_diag_server.c [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+		nrf_rpc_cbor_rsp_no_err(group, &rsp_ctx);                                          \
+	}                                                                                          \
+	NRF_RPC_CBOR_CMD_DECODER(ot_group, func_name, call_id, func_name, NULL)
+
+NRF_RPC_CBOR_CMD_STRING_GETTER(ot_get_version_string_getter, otGetVersionString,
+			       OT_RPC_CMD_GET_VERSION_STRING, char,);
+
+NRF_RPC_CBOR_CMD_STRING_GETTER(ot_get_extaddr, otLinkGetExtendedAddress,
+			       OT_RPC_CMD_LINK_GET_EXTENDED_ADDRESS, otExtAddress,
+			       openthread_get_default_context()->instance);
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/server/ot_rpc_diag_server.c:110]: fixed.

File: subsys/net/openthread/rpc/server/ot_rpc_diag_server.c (hunk 15 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/server/ot_rpc_diag_server.c:128]: fixed.

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+static otMeshLocalPrefix mesh_prefix;
+
+static void get_uint_t(ot_rpc_id id, void *result, size_t result_size)
+{
+	const size_t cbor_buffer_size = 0;
+	struct nrf_rpc_cbor_ctx ctx;
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, cbor_buffer_size);
+
+	nrf_rpc_cbor_cmd_rsp(&ot_group, id, &ctx);
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:33]: fixed.

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+
+	if (!zcbor_uint_decode(ctx.zs, result, result_size)) {
+		nrf_rpc_err(-EBADMSG, NRF_RPC_ERR_SRC_RECV, &ot_group, id, NRF_RPC_PACKET_TYPE_CMD);
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:36]: fixed.

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, cbor_buffer_size);
+
+	nrf_rpc_cbor_cmd_rsp(&ot_group, id, &ctx);
+
+	if (!zcbor_bstr_decode(ctx.zs, &zst)) {
+		nrf_rpc_err(-EBADMSG, NRF_RPC_ERR_SRC_RECV, &ot_group, id, NRF_RPC_PACKET_TYPE_CMD);
+		goto exit;
+	}
+
+	bytes_copied = buffer_size < zst.len ? buffer_size : zst.len;
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:58]: Please check the newer version.

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+{
+	return 0;
+}
+
+const otExtAddress *otLinkGetExtendedAddress(otInstance *aInstance)
+{
+	int ret_size = get_string(OT_RPC_CMD_LINK_GET_EXTENDED_ADDRESS, mac.m8, sizeof(mac.m8));
+
+	if (ret_size != sizeof(mac.m8)) {
+		LOG_ERR("Received mac addr size is too short");
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:97]: please check newer version.

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c (hunk 10 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/client/ot_rpc_diag_client.c]: fixed.

File: subsys/net/openthread/rpc/common/ot_rpc_ids.h (hunk 11 above)
Comment (alxelax) [reply subsys/net/openthread/rpc/common/ot_rpc_ids.h:18]: As I already mentioned in @Damian-Nordic's PR just usage `enum ot_rpc_cmd_server` is sufficient and this is according to Zephyr's coding paradigm (https://github.com/nrfconnect/sdk-nrf/pull/16406#discussion_r1673925000). Creation of new type is not required.

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:21]: Is it possible to get anyhow length of this name from build system?

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c (hunk 20 above)
[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:109]: I guess this is possible to use in all such places this function: https://github.com/nrfconnect/sdk-nrf/blob/collab-serial-ble-ot/subsys/net/openthread/rpc/common/ot_rpc_common.h#L21
but it can be done as final clean up step that we discussed today

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:21]: The definition looks like this: https://github.com/nrfconnect/sdk-openthread/blob/main/src/core/api/instance_api.cpp#L175 so I don't think so. However suggestions are welcomed.

[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:21]: I see, seems no way to minimize memory usage.

File: subsys/net/openthread/rpc/common/ot_rpc_ids.h (hunk 11 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/common/ot_rpc_ids.h]: Agreed I removed type def from enum.

File: subsys/net/openthread/rpc/client/ot_rpc_diag_client.c [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+		nrf_rpc_cbor_decoding_done(&ot_group, &ctx);
+		ot_rpc_report_decoding_error(id);
+		goto exit;
+	}
+
+	if (buffer_size >= zst.len) {
+		bytes_copied = zst.len;
+	} else {
+		bytes_copied = 0;
+		ot_rpc_report_decoding_error(id);
```
Comment (Damian-Nordic) [inline subsys/net/openthread/rpc/client/ot_rpc_diag_client.c:66]: `_decoding_done` is not called in this case. Call it before checking the length?

**Review by alxelax — APPROVED** [review]
